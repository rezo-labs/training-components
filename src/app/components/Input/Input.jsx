import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import warringSVG from './assets/warning.svg';
import successSVG from './assets/success.svg';
import errorSVG from './assets/error.svg';

const theme = {
    minimal: ['transparent', '#000000', 'white'],
    compact: ['', '#000000', 'white'],
    default: ['#cccccc', '#808080', 'white'],
    autofill: ['transparent', '#000000', '#e6ffe6'],
    focus: ['#0099ff', '#000000', 'white'],
};
const themeStatus = {
    error: ['#d93025', '#000000', 'white'],
    success: ['#00802b', '#000000', 'white'],
    warning: ['#ffc107', '#000000', 'white'],
};
const Myinput = styled.input`
    margin: 5px 0;
    padding: 15px;
    font-size:23px;
    width:350px;
    color: ${props => (props.isDisable ? '' : props.status ? themeStatus[props.status][1]
        : props.variant ? theme[props.variant][1] : theme.default[1])};
    border: 3px solid ${props => (props.isDisable ? 'transparent' : props.status ? themeStatus[props.status][0]
        : props.variant ? theme[props.variant][0] : theme.default[0])};
    background: ${props => (props.isDisable ? '' : props.status ? themeStatus[props.status][2]
        : props.variant ? theme[props.variant][2] : theme.default[2])};
    outline:none;
    border-radius:5px;
    :disabled{
        color: #b3b3b3;
    };
    :focus{
        border: 3px solid #0099ff;
        background: white;
        color:black;
    };
    :hover{
        color: #000000;
        background: #e1e1d0;
        border: 3px solid #cccccc;
    };
`;

const Mylabel = styled.label`
     display:block;
     color: #808080;
     font-size:18px;
`;
const Mydiv = styled.div`
    width:fit-content;
    font-family: Arial, Helvetica, sans-serif;
    margin-bottom:10px;
    font-weight:500;
`;
const Myimg = styled.img`
    margin-right:5px;
    object-fit:cover;
    width:20px;
    height:20px;
`;
const Helptext = styled.div`
    color: ${props => (props.status ? themeStatus[props.status][0] : theme.default[1])};  
    display:flex;
    line-height:20px;
`;
const Divtext = styled.span`
    font-size: 15px;
`;
export default function Input(props) {
    const {
        type, variant, isDisable, defaultValue, onChange,
        label, helpText, placeholder,
        id, onBlur, onMouseDown, onFocus, value, status,
    } = props;
    if (type === 'password' || type === 'text' || type === 'number') {
        return (
            <Mydiv>
                {label
                    ? (
                        <Mylabel>
                            {label}
                        </Mylabel>
                    ) : null }
                {isDisable
                    ? (
                        <Myinput
                            type={type}
                            disabled
                            id={id}
                            isDisable={isDisable}
                            onFocus={onFocus}
                            onBlur={onBlur}
                            onChange={onChange}
                            onMouseDown={onMouseDown}
                            value={value}
                            defaultValue={defaultValue}
                        />
                    )
                    : (
                        <Myinput
                            status={status}
                            type={type}
                            variant={variant}
                            placeholder={placeholder}
                            id={id}
                            onChange={onChange}
                            onFocus={onFocus}
                            onBlur={onBlur}
                            onMouseDown={onMouseDown}
                            value={value}
                            defaultValue={defaultValue}
                        />
                    )
                }
                {helpText ? (
                    <Helptext
                        status={status}
                    >
                        {status === 'warning'
                            ? (<Myimg src={warringSVG} />)
                            : status === 'success'
                                ? (<Myimg src={successSVG} />)
                                : status === 'error'
                                    ? (<Myimg src={errorSVG} />) : null
                        }
                        <Divtext>
                            {helpText}
                        </Divtext>
                    </Helptext>
                ) : null
                }
            </Mydiv>
        );
    }
}
Input.defaultProps = {
    variant: 'default',
    isDisable: false,
    onBlur: null,
    onChange: null,
    onMouseDown: null,
    onFocus: null,
    id: null,
    label: null,
    helpText: null,
    placeholder: null,
    defaultInput: '',
};
Input.propTypes = {
    /* Required value */
    /** User's Type */
    type: PropTypes.oneOf(['number', 'text', 'password']).isRequired,
    /* optional value */
    /** User's variant */
    variant: PropTypes.oneOf(['default', 'minimal', 'autofill', 'focus']),
    /** User's status */
    status: PropTypes.oneOf(['warning', 'error', 'success']),
    /** User's isDisable is optional but it's the first priority in render */
    isDisable: PropTypes.bool,
    /* String value */
    id: PropTypes.string,
    label: PropTypes.string,
    helpText: PropTypes.string,
    placeholder: PropTypes.string,
    /* Event propTypes: */
    /** Function called when user presses a mouse over the input  */
    onMouseDown: PropTypes.func,
    /** Function called when user leaves the input after focusing. */
    onBlur: PropTypes.func,
    /** Function called when the value of the input has been changed */
    onChange: PropTypes.func,
    /** Function called when the input get focus. */
    onFocus: PropTypes.func,
};
