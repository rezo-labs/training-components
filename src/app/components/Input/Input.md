# ---My Input
## 1. Input status:
```jsx
<Input status='warning' label = 'This is label' helpText='Help text here!' type = 'text'/>
<Input status='success' label = 'This is label' helpText='Help text here!' type = 'text'/>
<Input status='error' label = 'This is label' helpText='Help text here!' type = 'text'/>
```
## 2. variant:
```jsx
<Input variant ='default' label = 'default:' type = 'text'/>
<Input variant='autofill' label = 'autofill:' type = 'text'/>
<Input variant='minimal' label = 'minimal:' type = 'text'/>
<Input variant='focus' label = 'focus:' type = 'text'/>
```
## 3. isDisable:
```jsx
<Input isDisable label = 'This is label' helpText='help text here!' value='This is isDisable' type = 'text'/>
```
## 4. type:
```jsx
<Input label = 'Type password' helpText='help text here!' defaultValue='helloguys' type = 'password'/>
<Input label = 'Type number' helpText='help text here!' defaultValue='123321' type = 'number'/>
<Input label = 'Type text' helpText='help text here!' defaultValue='This is text type' type = 'text'/>
```
## 5. place holder:
```jsx
<Input variant ='focus' label = 'Type text' helpText='help text here!' placeholder='this is place holder' type = 'text'/>
```
## 6. check onBlur event:
```jsx
<Input status='warning' label = 'Type text' helpText='Help text here!' placeholder='this is place holder' type = 'text' onBlur={event => alert('you have left the input ...')}/>
```
## 7. check onChange event:
```jsx
<Input status='success' label = 'Type text' helpText='Help text here!' placeholder='this is place holder' type = 'text' onChange={ event => alert('you have Changed the input ...')  }/>
```
## 8. check onMouseDown event:
```jsx
<Input status='error' label = 'Type text' helpText='Help text here!' placeholder='this is place holder' type = 'text' onMouseDown={ event => alert('MouseDown event is started ')  }/>
```
## 9. check onFocus event:
```jsx
<Input status='warning' label = 'Type text' helpText='Help text here!' placeholder='this is place holder' type = 'text' onFocus={ event => alert('you have focused the input ...')  }/>
```


