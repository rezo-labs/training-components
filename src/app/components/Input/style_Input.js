import styled from 'styled-components';

const theme = {
    minimal: ['transparent', '#000000', 'white'],
    compact: ['', '#000000', 'white'],
    default: ['#cccccc', '#808080', 'white'],
    autofill: ['transparent', '#000000', '#e6ffe6'],
    focus: ['#0099ff', '#000000', 'white'],
};
const themeStatus = {
    error: ['#d93025', '#000000', 'white'],
    success: ['#00802b', '#000000', 'white'],
    warning: ['#ffc107', '#000000', 'white'],
};
export const Myinput = styled.input`
    margin: 5px 0;
    padding: 15px;
    font-size:23px;
    width:350px;
    color: ${props => (props.isDisable ? '' : props.status ? themeStatus[props.status][1]
        : props.variant ? theme[props.variant][1] : theme.default[1])};
    border: 3px solid ${props => (props.isDisable ? 'transparent' : props.status ? themeStatus[props.status][0]
        : props.variant ? theme[props.variant][0] : theme.default[0])};
    background: ${props => (props.isDisable ? '' : props.status ? themeStatus[props.status][2]
        : props.variant ? theme[props.variant][2] : theme.default[2])};
    outline:none;
    border-radius:5px;
    :disabled{
        color: #b3b3b3;
    };
    :focus{
        border: 3px solid #0099ff;
        background: white;
        color:black;
    };
    :hover{
        color: #000000;
        background: #e1e1d0;
        border: 3px solid #cccccc;
    };
`;

export const Mylabel = styled.label`
     display:block;
     color: #808080;
     font-size:18px;
`;
export const Mydiv = styled.div`
    width:fit-content;
    font-family: Arial, Helvetica, sans-serif;
    margin-bottom:10px;
    font-weight:500;
`;
export const Myimg = styled.img`
    margin-right:5px;
    object-fit:cover;
    width:20px;
    height:20px;
`;
export const Helptext = styled.div`
    color: ${props => (props.status ? themeStatus[props.status][0] : theme.default[1])};  
    display:flex;
    line-height:20px;
`;
export const Divtext = styled.span`
    font-size: 15px;
`;
