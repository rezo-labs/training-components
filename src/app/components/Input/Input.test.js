import React from 'react';
import { shallow, mount } from 'enzyme';

import Input from './Input';

describe('Input', () => {
    it('Render as expected', () => {
        const wrapper = shallow(<Input type="text" status="warning" />);
        expect(wrapper.find('*')).not.toHaveLength(0);
    });
    it('Check whether isDisable working', () => {
        const wrapper = shallow(<Input type="text" status="warning" variant="autofill" isDisable />);
        expect(wrapper.find('*')).not.toHaveLength(0);
    });
    it('Should throw errors', () => {
        console.error = jest.fn();
        shallow(<Input status="warning" variant="autofil" type="text" />);
        shallow(<Input status="waning" variant="autofill" type="text" />);
        expect(console.error).toHaveBeenCalledTimes(2);
    });
    it('check onChange callback', () => {
        const onChange = jest.fn();
        const testInputOnChange = mount(<Input type="text" onChange={onChange} />).find('input');
        testInputOnChange.simulate('change', { target: { value: 'this text is changed' } });
        expect(onChange).toHaveBeenCalled();
    });
    it('check onFocus callback', () => {
        const onFocus = jest.fn();
        const testInputOnFocus = mount(<Input type="text" onFocus={onFocus} />).find('input');
        testInputOnFocus.simulate('focus');
        expect(onFocus).toHaveBeenCalledTimes(1);
    });
    it('check onBlur callback', () => {
        const onBlur = jest.fn();
        const testInputOnBlur = mount(<Input type="text" onBlur={onBlur} />).find('input');
        testInputOnBlur.simulate('blur');
        expect(onBlur).toHaveBeenCalledTimes(1);
    });
    it('check onMouseDown callback', () => {
        const onMouseDown = jest.fn();
        const testInputOnMouseDown = mount(<Input type="text" onMouseDown={onMouseDown} />).find('input');
        testInputOnMouseDown.simulate('mousedown');
        expect(onMouseDown).toHaveBeenCalledTimes(1);
    });
});
