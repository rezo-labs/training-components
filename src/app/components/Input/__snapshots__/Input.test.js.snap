// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input snapshot testing 1`] = `
<div
  className="Input__Mydiv-b788xn-2 hNWqxr"
>
  <label
    className="Input__Mylabel-b788xn-1 hltnNq"
  >
    snapshot's using label
  </label>
  <input
    className="Input__Myinput-b788xn-0 kBbxPJ"
    id={null}
    onBlur={null}
    onChange={null}
    onFocus={null}
    onMouseDown={null}
    placeholder={null}
    type="text"
  />
  <div
    className="Input__Helptext-b788xn-4 fEoPYl"
  >
    <img
      className="Input__Myimg-b788xn-3 dWfZvU"
      src="test-file-stub"
    />
    <span
      className="Input__Divtext-b788xn-5 jkxnH"
    >
      helpText Snapshot testing
    </span>
  </div>
</div>
`;
